src/test/java/com/google/api/codegen/testdata/csharp_gapic_resourcenames_library.baseline

```diff
@@ -0,0 +1,719 @@
+        /// <param name="bookId">The book ID. Must not be <c>null</c>.</param>
+        /// <param name="returnId">The return ID. Must not be <c>null</c>.</param>
+        public ReturnName(string shelfId, string bookId, string returnId)
+        {
+            ShelfId = GaxPreconditions.CheckNotNull(shelfId, nameof(shelfId));
+            BookId = GaxPreconditions.CheckNotNull(bookId, nameof(bookId));
+            ReturnId = GaxPreconditions.CheckNotNull(returnId, nameof(returnId));
+        }
+
+        /// <summary>
+        /// The shelf ID. Never <c>null</c>.
+        /// </summary>
+        public string ShelfId { get; }
+
+        /// <summary>
+        /// The book ID. Never <c>null</c>.
+        /// </summary>
+        public string BookId { get; }
+
+        /// <summary>
+        /// The return ID. Never <c>null</c>.
+        /// </summary>
+        public string ReturnId { get; }
+
+        /// <inheritdoc />
+        public override string ToString() => s_template.Expand(ShelfId, BookId, ReturnId);
+    }
+
+    public sealed partial class BookNameOneof : IResourceName
+    {
+        public enum OneofType
+        {
+            BookName = 1,
+            ArchivedBookName = 2,
+            DeletedBookNameFixed = 3,
+        }
+
+        public static BookNameOneof Parse(string name)
+        {
+            BookNameOneof result;
+            if (TryParse(name, out result))
+            {
+                return result;
+            }
+            throw new ArgumentException("Invalid name", "name");
+        }
+
+        public static bool TryParse(string name, out BookNameOneof result)
+        {
+            BookName bookName;
+            if (BookName.TryParse(name, out bookName))
+            {
+                result = new BookNameOneof(OneofType.BookName, bookName);
+                return true;
+            }
+            ArchivedBookName archivedBookName;
+            if (ArchivedBookName.TryParse(name, out archivedBookName))
+            {
+                result = new BookNameOneof(OneofType.ArchivedBookName, archivedBookName);
+                return true;
+            }
+            DeletedBookNameFixed deletedBookNameFixed;
+            if (DeletedBookNameFixed.TryParse(name, out deletedBookNameFixed))
+            {
+                result = new BookNameOneof(OneofType.DeletedBookNameFixed, deletedBookNameFixed);
+                return true;
+            }
+            result = null;
+            return false;
+        }
+
+        public static BookNameOneof From(BookName bookName) => new BookNameOneof(OneofType.BookName, bookName);
+
+        public static BookNameOneof From(ArchivedBookName archivedBookName) => new BookNameOneof(OneofType.ArchivedBookName, archivedBookName);
+
+        public static BookNameOneof From(DeletedBookNameFixed deletedBookNameFixed) => new BookNameOneof(OneofType.DeletedBookNameFixed, deletedBookNameFixed);
+
+        public BookNameOneof(OneofType type, IResourceName value)
```

Comment:
This doesn't validate that the type we give is actually sane, of course. I may have mentioned elsewhere that we could check that.